### ISSUE TITLE
Misleading instructions for ComplexConv2D input shape?

### ISSUE THREAD

**Issue by shardie1992** [JesperDramsch/keras-complex#18]
Hi there,

First off I would just like to say thank you so much for this repo; it's been amazingly helpful!

That being said, I've run into a little snag when specifiying the input shape for a ComplexConv2D layer.

In the source code for this layer, the help section states the following (https://github.com/JesperDramsch/keras-complex/blob/master/complexnn/conv.py#L729-L733): 

"When using this layer as the first layer in a model,
    provide the keyword argument `input_shape`
    (tuple of integers, does not include the sample axis),
    e.g. `input_shape=(128, 128, 3)` for 128x128 RGB pictures
    in `data_format="channels_last"`."

Where I would naturally assume 128 = width, 128 = height, 3 = # of channels. Yet when I attempt to follow these instructions by executing the following code:

```
model = models.Sequential()

model.add(complexnn.conv.ComplexConv2D(32, (3, 3), activation='relu', padding='same', input_shape=(128, 128, 3)))
```
I am then met with the following error:

```
ValueError: Input 0 is incompatible with layer complex_conv2d_11: expected axis -1 of input shape to have value 2 but got shape (None, 128, 128, 3)
```

The layer consistently insists that the final axis be of value 2 and sure enough when I make input_shape=(128, 128, 2) it works just fine. Analogous versions of the same error/solution occur for ComplexConv1D and ComplexConv3D layers as well.

The insistence of the value 2 in the final axis makes me think it has something to do with splitting the input data up into real and imaginary parts into separate arrays of float dtypes. This is somewhat confusing to me though, as I was under the impression the input was meant to be specified as an array of (width, height, # of channels) where width and height were complex dtypes? If this isn't the case, and the data needs to be split into real and imaginary components, then how (if at all) is the number of channels specified in the ComplexConv2D layer?

In summary, I was just hoping for some clarification on the following:

1. Should the input data be transformed into separate arrays of real and imaginary float dtypes? Or can it be kept as a single array of complex dtypes?

2. How exactly should the input_shape be specified (particularly when it comes to the number of channels the data has) ?

My apologies if I'm missing something trivial or if this has been answered before!

**Comment by JesperDramsch** [comment]
Hi @shardie1992,

thank you very much for the detailed question. I apologize the docstring is wrong and confusing. This is a remnant from working off the keras docstring.

1. Should the input data be transformed into separate arrays of real and imaginary float dtypes? Or can it be kept as a single array of complex types?

Currently it has to be separate real and imaginary parts as `float`. This is shown in the [Documentation](https://keras-complex.readthedocs.io/math.html), but re-reading the documentation and the readme here, this should absolutely be clearer. Apologies.

How exactly should the input_shape be specified (particularly when it comes to the number of channels the data has) ?

Multiples of two, first all the real-valued data, then the corresponding complex-valued values [real1 real2 real3 real4 complex1 complex2 complex3 complex4]
Refer to https://github.com/JesperDramsch/keras-complex/blob/dc4b986acc69cbe7351e475882ed269381d6f594/complexnn/conv.py#L377

I hope this helps. This definitely needs an improvement, if you have the time to add to the documentation that would be highly appreciated.

**Comment by shardie1992** [comment]
No worries! I had a feeling it might have been a remnant of the keras docstring so that's completely understandable! 

Regarding your instructions to organize the complex data as [real1 real2 real3 real4 imag1 imag2 imag3 imag4]... After posting my question, I was looking over some of the closed issues on this repo and came across a similar question that had been posed before: https://github.com/JesperDramsch/keras-complex/issues/10

In that issue you actually suggest to organize the complex data as [real1 imag1 real2 imag2 real3 imag3 real4 imag4]... This is at odds with your earlier instructions so I was just hoping to make extra sure which form is the correct one?

Finally, I would gladly add to the documentation to make all this a little clearer! Unfortunately, I'm quite new to github in general so any advice on how to do so would be greatly appreciated :P

**Comment by JesperDramsch** [comment]
Oh no. That was a blunder on my part.

https://github.com/JesperDramsch/keras-complex/blob/dc4b986acc69cbe7351e475882ed269381d6f594/complexnn/conv.py#L378
https://github.com/JesperDramsch/keras-complex/blob/dc4b986acc69cbe7351e475882ed269381d6f594/complexnn/conv.py#L379

show how it's split [r,r,r,i,i,i]. I'll correct that in #10 .

**Comment by shardie1992** [comment]
Ok great, thanks for clearing that up! 

As a quick side note; in my quest to personally determine which form was the correct one, I came across what is seemingly an error in the code here: 

https://github.com/JesperDramsch/keras-complex/blob/dc4b986acc69cbe7351e475882ed269381d6f594/complexnn/conv.py#L491-L496

The brackets around "2 * input_dim, 2 * self.filters" seem to be missing in line 496. Correct me if I'm wrong but the way the if statement is constructed seems to ensure that the else block is executed everytime? Strangely though, I am not getting an error when calling ComplexConv, yet when I attempt to execute similarly formatted code in isolation, errors do occur. Not sure if there's some exception to the syntax going over my head here but I thought I'd let you know just in case. 

I would have added this (and the earlier shape clarifications) to the documentation directly but unfortunately my experience with github is still quite limited. I plan on using this repo quite a lot in the coming months, so if the upward trajectory of my github experience continues then I'll gladly do so in the future!

In any case, thanks again for all your assistance!

**Comment by King-Of-Knights** [comment]
@shardie1992 I am facing the same issue as you. I am researching the network pruning techniques but I found some strange behavior of input dim. Like, I got output dim = 32 for the 1st but got 16 input dim for the next layer without pool layer and stride in convoluted layer. Maybe the missing of bracket is behind of all of this. I will check it and talk about it later

**Comment by jintaoAndshan** [comment]
hello，author，do you have the complexnn of tf 2.x version？
